### PR TITLE
feat(input-chips): new module

### LINKED ISSUE #2159: New component: input chips
Original issue https://github.com/eBay/skin/issues/2108
This will build the input with the chip module that we built from that last issue.

### PR DESCRIPTION
**Pull request by ArtBlue**
Fixes #2159 

- [X] This PR contains CSS changes
- [ ] This PR does not contain CSS changes

## Description
Created new input chips module that includes chips and a combobox selector.

## Notes
WIP

## Screenshots
Basic
image

With Label
image

With Helper Text
image

Disabled State
image

Error State
image

Expanded
image

## Checklist

- [x] I verify the build is in a non-broken state
- [x] I verify all changes are within scope of the linked issue

- [x] I regenerated all CSS files under dist folder
- [x] I tested the UI in all supported browsers
- [ ] I did a visual regression check of the components impacted by doing a Percy build and approved the build
- [x] I tested the UI in dark mode and RTL mode
- [x] I added/updated/removed Storybook coverage as appropriate

### CODE REVIEW

File: docs/_includes/input-chips.html
Comment (agliga) [inline docs/_includes/input-chips.html:13]: Is there a reason why we have 3 levels? Can we not simply flatten this by one level?

So like all styles for `.input-chips__container` becomes `.input-chips` style. Then we could rename `input-chips__content` to be `input-chips__container` (I prefer the name input chips container for the container of input chips)

File: docs/_includes/input-chips.html
Comment (agliga) [inline docs/_includes/input-chips.html:150]: Do we need this extra layer? It seems like we can wrap that in the parent as well.

**Review by agliga — COMMENTED** [review]
Looks good. I think we could probably flatten some of this structure. It seems like there are too many levels and Im wondering if we can make it much simpler that way.

File: docs/_includes/input-chips.html
Comment (ArtBlue) [reply docs/_includes/input-chips.html:150]: I'll take a look.

Comment (ArtBlue) [reply docs/_includes/input-chips.html:150]: Actually, `input-chips_items` is necessary since there's a flex there - the group of chips have to be grouped and the toggle button has to be separate.

File: docs/_includes/input-chips.html
Comment (ArtBlue) [reply docs/_includes/input-chips.html:13]: Ok, so the other reason why there's that extra layer is because `input-chips` has an attached label and helper text that both need to sit outside the box.

File: docs/_includes/input-chips.html
Comment (ArtBlue) [reply docs/_includes/input-chips.html:150]: This is what happens we we flatten. 
image

**Comment by LuLaValva** [comment]
It's a little bit weird that you need to click on the text section to get the combobox dropdown; I'm not sure if this would cause accessibility concerns, but as a user I'd expect that clicking anywhere inside the box _besides_ on a chip would bring up the combobox and allow me to start typing.

**Comment by ArtBlue** [comment]
> It's a little bit weird that you need to click on the text section to get the combobox dropdown; I'm not sure if this would cause accessibility concerns, but as a user I'd expect that clicking anywhere inside the box _besides_ on a chip would bring up the combobox and allow me to start typing.

We actually talked about that previously. I don't see any a11y issues with that. In fact, trying to mimic that for screen readers might be the thing that causes an issue because to trigger that with a keyboard and screen reader, you'd need the chip container to get focus and be interactive in and of itself - as far as I know that's more likely to be problematic. 

Visually speaking, I'm not convinced there's enough affordance there to *expect* a click anywhere inside the chip container to focus on the input and toggle the combobox. The reason why we insisted the design to have the placeholder text was actually for that very reason - not enough visual affordance and expectation of "turning on" the combobox when you click into the box. I suppose *some users* might expect that, but that would most likely be based on whether they've interacted with similar components. IMO this being far from comparable to any native component, I'm not sure that's a universal expectation. In fact, in certain situations (when there are enough chips to cover most of the box) doing that would not make a difference. We could definitely see if we can do some user testing and see if that's a gap.

That said, that *could be* a nice *layered* feature. Technically, it's another additive option for discoverability, which might be fine, but I doubt the lack of having it is an actual issue. My only major concern would be since the chip close buttons are not that big, someone missing the close target would prompt to open the combobox - not a major unrecoverable issue, but still...

Also, I'm not sure to what extent we should be building the interaction into Skin itself. IMO, as long as the interaction allows for all the UIs to surface for the docs to make visual sense (it does currently), Skin doesn't necessarily need to have the exhaustive set of interaction use cases. Those things should probably live in Ebay UI core.

**Comment by LuLaValva** [comment]
> I'm not convinced there's enough affordance there to expect a click anywhere inside the chip container to focus on the input and toggle the combobox.

I agree this is a good progressive enhancement to add to eBayUI which doesn't necessarily need to be in `skin`, except for one exception. Right now clicking on the chevron on the top right has no effect, which was confusing for me so it will probably also be confusing for users

**Comment by ArtBlue** [comment]
> I agree this is a good progressive enhancement to add to eBayUI which doesn't necessarily need to be in `skin`, except for one exception. Right now clicking on the chevron on the top right has no effect, which was confusing for me so it will probably also be confusing for users

Good catch. I was meaning to hook this up, but forgot. I'll add that.

**Comment by ArtBlue** [comment]
Ok, I hooked up the combobox expansion toggle on button click. One very important thing that I haven't finalized yet...

❗ IMPORTANT 

Right now, the combobox list interaction is isolated to the input. I've been strongly considering decoupling that (removing `combobox` component classes?) and instead building that into `.input-chips`. 

As part of my last change, I added `.input-chips--expanded` to allow cascading visual updates based on state to support that. However, right now the combobox list is still isolated and not integrated into that same cascading flow. I'm not sure what type of `a11y` implications a change like this might have, but it seem like overall the correct path forward.

File: src/less/input-chips/input-chips.less
Comment (saiponnada) [inline src/less/input-chips/input-chips.less:9]: these need to be moved to CSS variables

File: docs/_includes/input-chips.html
Comment (saiponnada) [inline docs/_includes/input-chips.html:40]: - This button needs to have appropriate label to convey to SR users. 
- Also, I noticed chevron not reflecting appropriately when focus in exited using keyboard i.e chevron points up even though the listbox is closed.
- The input is properly conveying that a popup listbox is associated. Do we also need to convey expanded/collapsed state on the chevron button?

File: src/less/input-chips/input-chips.less
Comment (ArtBlue) [reply src/less/input-chips/input-chips.less]: done!

File: docs/_includes/input-chips.html
Comment (ArtBlue) [reply docs/_includes/input-chips.html]: > This button needs to have appropriate label to convey to SR users.

Good catch! Done.

> Also, I noticed chevron not reflecting appropriately when focus in exited using keyboard i.e chevron points up even though the listbox is closed.

That's a known issue. This is an interaction piece that is more appropriate to handle in a js library, in our case, ebay ui core.

> The input is properly conveying that a popup listbox is associated. Do we also need to convey expanded/collapsed state on the chevron button?

Done. Let me know how that looks. I tried to mimic the same attributes.

File: src/less/input-chips/input-chips.less
Comment (LuLaValva) [inline src/less/input-chips/input-chips.less:21]:
```suggestion
.input-chips__items {
```

File: src/less/input-chips/input-chips.less
Comment (LuLaValva) [inline src/less/input-chips/input-chips.less:77]:
```suggestion
.input-chips--expanded .input-chips__list-toggle {
```

**Review by LuLaValva — COMMENTED** [review]
Is it proper BEM to use a single underscore?

Otherwise, looks awesome!

File: src/less/input-chips/input-chips.less
Comment (LuLaValva) [inline src/less/input-chips/input-chips.less:65]: Should we really be deviating this much from core listbox styles? I would imagine that the only part we should override is the positioning.

File: docs/_includes/input-chips.html
Comment (ianmcburnie) [reply docs/_includes/input-chips.html:13]: Hmm, I think we should look at keeping the label outside the scope of this module, like we do for textbox and combobox. That way developers can position the label wherever they like. Unless I'm missing something?

File: docs/_includes/input-chips.html
Comment (ianmcburnie) [inline docs/_includes/input-chips.html:40]: I'm not sure why we have a button here. This is a deviation from combobox pattern.

File: docs/_includes/input-chips.html
Comment (ianmcburnie) [inline docs/_includes/input-chips.html:141]: Similar to the label comment above, we might want to consider removing any responsibility for helper text placement too, and have this module responsible only for the layout of elements that sit within the combobox boundary.

**Comment by ianmcburnie** [comment]
I still think there may be a better name than `input-chips`. I liked what we did for `date-textbox`, as the name lets us know it is a textbox specifically for dates. I wonder if we can do something similar here and incorporate the underlying control (a combobox) in the name...?

**Comment by ianmcburnie** [comment]
> Is it proper BEM to use a single underscore?

No. Should be double underscore.

File: docs/_includes/input-chips.html
Comment (agliga) [reply docs/_includes/input-chips.html:13]: Agreed. Field should be used in that case.

File: docs/_includes/input-chips.html
Comment (agliga) [reply docs/_includes/input-chips.html]: I mean is this without changing any styles? 
The way I see it is if we do:
input-chips root container (should display flex, column I think)
    input-chips-container (contains all chips, display flex, row)
    combobox input
I would think that should work fine.

**Comment by ianmcburnie** [comment]
FYI looks like a couple of stale dist files crept in after I build. I wonder if prettier config is not picking these up when staged.

Screenshot 2023-11-17 at 11 22 16 AM

File: src/less/input-chips/input-chips.less
Comment (ArtBlue) [reply src/less/input-chips/input-chips.less]: If DS has the current specs where they have it, and we're trying to make it look how they want, which is a purely visual "preference," should the fact that we've chosen to use combobox under the hood be enough to overturn the visual specs that DS has landed on?

File: docs/_includes/input-chips.html
Comment (ArtBlue) [reply docs/_includes/input-chips.html]: @ianmcburnie , I tried numerous ways to retain the toggle button inside the combobox markup, but found it impossible. I can poke around for a way to do it. If possible, it will almost certainly come with a hefty amount of style overrides.

File: src/less/input-chips/input-chips.less
Comment (ArtBlue) [reply src/less/input-chips/input-chips.less]: Good catches!

File: docs/_includes/input-chips.html
Comment (ianmcburnie) [reply docs/_includes/input-chips.html]: I'm saying we don't need a button at all. The combobox just expands on focus doesn't it?

Comment (ArtBlue) [reply docs/_includes/input-chips.html]: I recall we fought tooth and nail with DS to retain the toggle button during all the design iterations because of its superior UX of providing an immediate visual indication of available options to choose from. If we rely solely on focus to expose the options, we're taking a step backwards because we're losing the immediate visual cue without the need for user action/discovery.

Comment (ianmcburnie) [reply docs/_includes/input-chips.html]: I meant we can keep the chevron, it just doesn't need to be in a button.

File: src/less/chips-combobox/chips-combobox.less
Comment (LuLaValva) [inline src/less/chips-combobox/chips-combobox.less:47]: I think this vertical flip is awesome, but we don't do it anywhere else so we should probably ask design about it first. Regular combobox also uses `rotate` instead of `rotateX` so it would be a spin instead of a flip if we added `transition`.

File: dist/tokens/evo-dark.css
Comment (LuLaValva) [inline dist/tokens/evo-dark.css:12]: It doesn't look like any of the changes in `evo-dark` or `evo-light` are used anywhere, I know they're probably for a future PR where we overhaul the dropdown but for now I think they should be removed since they are unused.

File: docs/_includes/chips-combobox.html
Comment (LuLaValva) [inline docs/_includes/chips-combobox.html:40]: All of these `TBD` blocks should be resolved before merging.

**Review by LuLaValva — COMMENTED** [review]
Looks great! Just a few minor comments.

File: dist/tokens/evo-dark.css
Comment (ArtBlue) [reply dist/tokens/evo-dark.css:12]: Good catch. I had removed the combobox overrides and forgot to remove these.

File: src/less/chips-combobox/chips-combobox.less
Comment (ArtBlue) [reply src/less/chips-combobox/chips-combobox.less:47]: The regular combobox actually uses 12px chevron. I'm assuming this will be easily swappable in ebay ui to the 16px chevron that's being used here and per DSs specs. The `css` transforms for combobox are specific to the 12px icon so it would not interfere with this.

I'd actually prefer to keep this, if possible. We can roll it out, demo for design, and if they don't like it, removing the transform and transition is easily doable without a breaking change.

**Comment by ianmcburnie** [comment]
Is there a reason we have deviated away from the default textbox width?

Screenshot 2023-12-18 at 11 38 32 AM
Screenshot 2023-12-18 at 11 38 23 AM
Screenshot 2023-12-18 at 11 38 17 AM

File: docs/_includes/chips-combobox.html
Comment (ianmcburnie) [inline docs/_includes/chips-combobox.html:194]: Looks like these are missing the disabled attribute as I can still focus on them:

Screenshot 2023-12-18 at 11 47 35 AM

Comment (ianmcburnie) [reply docs/_includes/chips-combobox.html:194]: I'm able to focus on the textbox portion too.

File: docs/_includes/chips-combobox.html
Comment (ianmcburnie) [inline docs/_includes/chips-combobox.html:14]: `chips-combobox-1-heading` doesn't seem to resolve to an id. Do we need aria-describedby and role=group here?

File: docs/_includes/chips-combobox.html
Comment (ianmcburnie) [inline docs/_includes/chips-combobox.html:13]: One thing I notice different from textbox and combobox is that they default to span tag (for easy inline/side-by-side layout,  just like a regular textbox) and here we are adopting div/block level layout. Should we consider following same approach as textbox & combobox, or are their fundamental differences here?

File: docs/_includes/chips-combobox.html
Comment (ArtBlue) [reply docs/_includes/chips-combobox.html:14]: ### aria-describedby
I thought I had commented on the heading IDs in the PR somewhere, but I lost track of where myself. 😄 

One of the PR feedbacks was to remove labels, helper text, and error messaging. When removing the label, I thought leaving the IDs in would be helpful to indicate for ebay ui to create an option in the component for teams to pass this ID in and have the combobox items linked to the label external to the module.  If it's more confusing in isolation than helpful, I can remove it. 

### role="group"
Don't we need to have the element containing all the chips be a group to indicate a group of `chip` items?

File: docs/_includes/chips-combobox.html
Comment (ArtBlue) [reply docs/_includes/chips-combobox.html:13]: Technically, I suppose it's possible for `chips-combobox` to be displayed inline on the same line as other form elements, but would we ever have `chips-combobox` components side by side with other form elements or each other? I don't personally see happening since `chips-combobox` is much larger than the typical form element. 

It's the main reason I used a div here. That said, if we want the flexibility anyway, I can go ahead and use spans instead.

Comment (ianmcburnie) [reply docs/_includes/chips-combobox.html:13]: I think it would be likely to be placed alongside others on desktop, especially menu-buttons. However, I think the fact that this is more of a composite component (due to it containing button chips) could lend weight to it operating differently from those other two. Same goes for the decision on 100% width really.  If it's a simple change to make it inline-block by default then maybe make the change for consistency sake with the other form fields we have built. Otherwise, maybe not a big deal if it stays the same (it just means devs will need additional css (and possibly an override) if they want things to line up side by side with it)

**Comment by ArtBlue** [comment]
> Is there a reason we have deviated away from the default textbox width?

I don't think the Figma had any min/max/default widths. I can't verify it at the moment (getting a blank page ATM).

From the dev perspective, since this is a composite component, it made sense to have it expand into the container limitations and allow implementation-level layouts to control its dimensions.

I suppose, we can do what we do with textbox and the fluid modifier. I'll look into that.

File: docs/_includes/chips-combobox.html
Comment (ianmcburnie) [reply docs/_includes/chips-combobox.html:14]: A few options:

1. No ARIA. Remove group and aria-describedby and leverage a simple unordered list (with no label/description) and hope that user can deduce the relationship to the combobox below it.
2. ARIA. Keep group and add `aria-label="Selected sports"` or a more generic `aria-label="Selected values". We could also add `aria-roledescription="Chips"` if we think chips might be a user interface term that users are familiar with.
3. A mix. Use unordered list (no role=group) and apply`aria-label="Selected sports"` to the ul

File: docs/_includes/chips-combobox.html
Comment (ianmcburnie) [inline docs/_includes/chips-combobox.html:109]: Being picky here, but Baseball is already selected as a chip (and so shouldn't be visible as a combobox option), so let's replace it with another sport.

Screenshot 2023-12-18 at 2 03 10 PM

**Review by agliga — COMMENTED** [review]
Looks great overall. I think its really come a long way.
One comment, is there a way to make this interactive on the docs? We have other components which are interactive like combobox and select. It really adds to the whole feel if a user can actually click an item in chips combobox and see it change. Or hit x and see it removed. It doesn't have to be complex, but at least to see the interactions.

File: docs/_includes/chips-combobox.html
Comment (ArtBlue) [reply docs/_includes/chips-combobox.html:13]: Ok, I'll go ahead and make the change.

**Review by ianmcburnie — APPROVED** [review]
Nice job @ArtBlue And thanks for working with design to remove those early a11y barriers.

**Comment by ianmcburnie** [comment]
@ArtBlue Oops, just spotted what looks like a couple of unsynced dist files in your branch:

Screenshot 2023-12-19 at 3 02 19 PM

**Comment by agliga** [comment]
There are some errors when I run this locally
```
> cat lint.log

/Users/agliga/build/ebay/skin/docs/src/js/main.js
  549:21  error  'sItemSelected' is never reassigned. Use 'const' instead   prefer-const
  559:21  error  'sChipText' is never reassigned. Use 'const' instead       prefer-const
  560:21  error  'elChipParentLI' is never reassigned. Use 'const' instead  prefer-const

✖ 3 problems (3 errors, 0 warnings)
  3 errors and 0 warnings potentially fixable with the `--fix` option.
```
Also I see a lot of files being deleted when I run locally. Not sure if its because of merge or not.

**Comment by ArtBlue** [comment]
> There are some errors when I run this locally
> 
> ```
> > cat lint.log
> 
> /Users/agliga/build/ebay/skin/docs/src/js/main.js
>   549:21  error  'sItemSelected' is never reassigned. Use 'const' instead   prefer-const
>   559:21  error  'sChipText' is never reassigned. Use 'const' instead       prefer-const
>   560:21  error  'elChipParentLI' is never reassigned. Use 'const' instead  prefer-const
> 
> ✖ 3 problems (3 errors, 0 warnings)
>   3 errors and 0 warnings potentially fixable with the `--fix` option.
> ```
> 

Wow! That is extremely opinionated for docs js. LOL. Ok, I'll change them.

> Also I see a lot of files being deleted when I run locally. Not sure if its because of merge or not.

I'm not sure. The diff on the PR doesn't have that. Maybe it's something local?

**Review by agliga — APPROVED** [review]
Lets wait for the floating ui PR to be merged before merging this.

**Pull request merged**
